lib/dal/src/workspace_snapshot/node_weight/action_node_weight.rs

```diff
@@ -211,3 +221,101 @@ impl CorrectTransforms for ActionNodeWeight {
+        // the prototype. The generic exclusive edge logic assumes there can be
+        // one and only one edge of the given kind, so we have to do a custom
+        // implementation here for actions, since they should have just one use
+        // edge to a component and one use edge to a prototype. This is simpler
+        // than rewriting all the graphs to have distinct edge kinds for action
+        // prototypes and/or the components the action is for.
+
+        struct ActionUseTargets {
+            component: Option<SplitGraphNodeId>,
+            prototype: Option<SplitGraphNodeId>,
+        }
+
+        let mut new_action_use_targets = ActionUseTargets {
+            component: None,
+            prototype: None,
+        };
+
+        let mut removal_set = BTreeSet::new();
+
+        for update in &updates {
+            match update {
+                si_split_graph::Update::NewEdge { destination, .. }
+                    if update.source_has_id(self.id())
+                        && update.is_of_custom_edge_kind(EdgeWeightKindDiscriminants::Use) =>
+                {
+                    removal_set.remove(&destination.id);
+                    match destination.custom_kind {
+                        Some(NodeWeightDiscriminants::ActionPrototype) => {
+                            new_action_use_targets.prototype = Some(destination.id);
+                        }
+                        Some(NodeWeightDiscriminants::Component) => {
+                            new_action_use_targets.component = Some(destination.id);
+                        }
+                        _ => {}
+                    }
+                }
+                si_split_graph::Update::RemoveEdge { destination, .. }
+                    if update.source_has_id(self.id())
+                        && update.is_of_custom_edge_kind(EdgeWeightKindDiscriminants::Use) =>
+                {
+                    removal_set.insert(&destination.id);
+                }
+                _ => {}
+            }
+        }
+
+        let removals: Vec<_> = graph
+            .outgoing_edges(self.id(), EdgeWeightKindDiscriminants::Use)?
+            .filter_map(|edge_ref| {
+                graph
+                    .node_weight(edge_ref.target())
+                    .and_then(|destination_weight| {
+                        match destination_weight {
+                            NodeWeight::ActionPrototype(_) => new_action_use_targets.prototype,
+                            NodeWeight::Component(_) => new_action_use_targets.component,
+                            _ => None,
+                        }
+                        .is_some_and(|new_destination_id| {
+                            new_destination_id != edge_ref.target()
+                                && !removal_set.contains(&new_destination_id)
+                        })
+                        .then_some(edge_ref.triplet())
+                    })
+            })
+            .collect();
```

Comment:
Worth noting, you don't need to collect into a Vec here--nobody's borrowing anything :)